underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AOutputStream.java

```diff
@@ -0,0 +1,152 @@
+  private final String mKey;
+
+  /** The local file that will be uploaded when the stream is closed. */
+  private final File mFile;
+
+  /** Flag to indicate this stream has been closed, to ensure close is only done once. */
+  private final AtomicBoolean mClosed = new AtomicBoolean(false);
+
+  /**
+   * A {@link TransferManager} to upload the file to S3 using Multipart Uploads. Multipart Uploads
+   * involves uploading an object's data in parts instead of all at once, which can work around S3's
+   * limit of 5GB on a single Object PUT operation.
+   *
+   * It is recommended (http://docs.aws.amazon.com/AmazonS3/latest/dev/UploadingObjects.html)
+   * to upload file larger than 100MB using Multipart Uploads.
+   */
+  private final TransferManager mManager;
+
+  /** The output stream to a local file where the file will be buffered until closed. */
+  private OutputStream mLocalOutputStream;
+
+  /** The MD5 hash of the file. */
+  private MessageDigest mHash;
+
+  /**
+   * Constructs a new stream for writing a file.
+   *
+   * @param bucketName the name of the bucket
+   * @param key the key of the file
+   * @param manager the transfer manager to upload the file with
+   * @throws IOException when a non-Alluxio related error occurs
+   */
+  public S3AOutputStream(String bucketName, String key, TransferManager manager)
+      throws IOException {
+    Preconditions.checkArgument(bucketName != null && !bucketName.isEmpty(), "Bucket name must "
+        + "not be null or empty.");
+    mBucketName = bucketName;
+    mKey = key;
+    mManager = manager;
+    mFile = new File(PathUtils.concatPath("/tmp", UUID.randomUUID()));
+    try {
+      mHash = MessageDigest.getInstance("MD5");
+      mLocalOutputStream =
+          new BufferedOutputStream(new DigestOutputStream(new FileOutputStream(mFile), mHash));
+    } catch (NoSuchAlgorithmException e) {
+      LOG.warn("Algorithm not available for MD5 hash.", e);
+      mHash = null;
+      mLocalOutputStream = new BufferedOutputStream(new FileOutputStream(mFile));
+    }
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    mLocalOutputStream.write(b);
+  }
+
+  @Override
+  public void write(byte[] b) throws IOException {
+    mLocalOutputStream.write(b, 0, b.length);
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    mLocalOutputStream.write(b, off, len);
+  }
+
+  @Override
+  public void flush() throws IOException {
+    mLocalOutputStream.flush();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed.getAndSet(true)) {
+      return;
+    }
+    mLocalOutputStream.close();
+    try {
+      ObjectMetadata meta = new ObjectMetadata();
+      if (SSE_ENABLED) {
+        meta.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
+      }
+      meta.setContentLength(mFile.length());
+      meta.setContentEncoding(Mimetypes.MIMETYPE_OCTET_STREAM);
```

Comment:
nit: maybe add a comment here to explain what's the magic MIMETYPE_OCTET_STREAM?